### ISSUE TITLE
Add support for --flat-classpath, --auto-export-maven-dependencies and --overrides

### ISSUE THREAD

**Issue by FroMage** [ceylon/ceylon-ide-eclipse#1291]
As recently added in the JVM compiler and runners. Note that `--overrides` is also supported for the JS compiler now.

**Comment by davidfestal** [comment]
I assume we should also enrich CeylonProjectConfig with these options if they are taken in account inside the Ceylon config file ?

**Comment by davidfestal** [comment]
About the `--flat-classpath` or `--auto-export-maven-dependencies` option, does it make sense to have them set in a way at compile-time, and differently at runtime ?
I'm inclined to add a single option in a project for both compilation and run, for the sake of simplicity, but I don't know if that would be too limited.

By the way, what about adding these 2 options in the Ceylon config file also ? It seems they are not for the moment (only the `overrides` option is).

@FroMage, @quintesse  : wdyt ?

**Comment by FroMage** [comment]
Yes they need to end up in the config file. About compile/runtime, I think we can start with both at once.

**Comment by davidfestal** [comment]
OK, Thanks for your answer !

**Comment by davidfestal** [comment]
Do you have a xsd or dtd for this format ?

**Comment by davidfestal** [comment]
it could be interesting to add it to the IDE so that its construction inside Eclipse is guided automatically by the XML plugin, and also validated.

**Comment by FroMage** [comment]
No I don't.

**Comment by FroMage** [comment]
But I agree it would be nice to have.

**Comment by davidfestal** [comment]
It seems to me it is implemented (at least as I understood the whole thing).
@FroMage : could you confirm that it works as you expected ?

**Comment by davidfestal** [comment]
You will find the new options in the Module Repositories project properties page.

**Comment by davidfestal** [comment]
However, in the IDE `Ceylon Module Dependencies` View of a project with an override for the import of a wrong version of `ceylon.collection`, I observed that the module name and version in the Ceylon global model (ie the Module object I suppose) is still the name and version of the originally imported module.
And the overriden module is now seen as a Java module, and not a Ceylon module anymore. Is there any reason for this ?

**Comment by davidfestal** [comment]
@gavinking : I hope that this possible difference between the name / version of a module (as contained in the `Module` object), and the equivalent information in the corresponding archive full path, would not introduce regressions in some places of the IDE, where we would have used the module name and version to search for paths...

**Comment by gavinking** [comment]
@davidfestal when I enter nonsense in the Overrides XML file field, I get an error message box popup complaining that the build failed. I think the preferences page needs to do better validation on the contents of that field.

**Comment by gavinking** [comment]
> I observed that the module name and version in the Ceylon global model (ie the Module object I suppose) is still the name and version of the originally imported module.
And the overriden module is now seen as a Java module, and not a Ceylon module anymore. Is there any reason for this ?

@FroMage?

**Comment by davidfestal** [comment]
@gavinking : I think that the best would be to write and XSD schema for the overrides.xml content. Which would allow :
- simplifying its authoring inside the Eclipse XML editor
- easy validating when selecting it.

**Comment by davidfestal** [comment]
About an overridden Ceylon module being seen as a Java module,
it's because, when loading the module descriptor from binaries, the model loader finds (considering it as an error) that **the version inside the `Module` class file is not the same as the version of the model `Module` object** (created according the the encountered module import).

Within the code, it's located [here](https://github.com/ceylon/ceylon-compiler/blob/82d658c746dc291d5e8cb96f49b4588215d5c05f/src/com/redhat/ceylon/compiler/loader/AbstractModelLoader.java#L1769).

Shouldn't there _be some information stored on the `Module` object_ to indicate that the original version (and possibly the name) were replaced by different values ?
So here we could compare with the real name and version, and not the name and version coming from the Module Import.

This would also allow showing the overrides in the dependency graph (into parenthesis for example, or with a specific marker associated to more precise information in the tooltip).

**Comment by davidfestal** [comment]
One of the side-effects of an overriden Ceylon module being seen as a Java module, is that we don't check for binary compatibility for it anymore (see [here](https://github.com/ceylon/ceylon-compiler/blob/e36ffde7c3d023e3c7ff7b8078d57a865abf5dc0/src/com/redhat/ceylon/compiler/loader/model/LazyModuleManager.java#L146)).

**Comment by davidfestal** [comment]
Also [this line](https://github.com/ceylon/ceylon-compiler/blob/82d658c746dc291d5e8cb96f49b4588215d5c05f/src/com/redhat/ceylon/compiler/loader/AbstractModelLoader.java#L1604) seems to show that the packages inside the Ceylon overriden module will all be considered as shared (just as in a real Java module).

**Comment by davidfestal** [comment]
After some research, the following lines might introduce bugs if the name of an overridden Ceylon Module is not the same as its original name :
- [`JDTModelLoader.java`, line 1140](https://github.com/ceylon/ceylon-ide-eclipse/blob/fb81b6b1224109262258d9fdb635864c18b2c781/plugins/com.redhat.ceylon.eclipse.ui/src/com/redhat/ceylon/eclipse/core/model/JDTModelLoader.java#L1140),
- [`Module.java`, line 181](https://github.com/ceylon/ceylon-spec/blob/9bb4136964e1aa1c2a94647921c8d38e9753023a/src/com/redhat/ceylon/compiler/typechecker/model/Module.java#L181),
- [`JDTModule.java`, line 958](https://github.com/ceylon/ceylon-ide-eclipse/blob/fb81b6b1224109262258d9fdb635864c18b2c781/plugins/com.redhat.ceylon.eclipse.ui/src/com/redhat/ceylon/eclipse/core/model/JDTModule.java#L958),
- [`FindDeclarationNodeVisitor.java`, line 47](https://github.com/ceylon/ceylon-ide-eclipse/blob/fb81b6b1224109262258d9fdb635864c18b2c781/plugins/com.redhat.ceylon.eclipse.ui/src/com/redhat/ceylon/eclipse/util/FindDeclarationNodeVisitor.java#L47),
- [`JavaSearch.java`, line 662](https://github.com/ceylon/ceylon-ide-eclipse/blob/7c4ccdca7d70fbf3d249aba4de377ec2a2fa1b98/plugins/com.redhat.ceylon.eclipse.ui/src/com/redhat/ceylon/eclipse/util/JavaSearch.java#L662),
- [`CeylonNavigatorContentProvider.java`, line 139](https://github.com/ceylon/ceylon-ide-eclipse/blob/7c4ccdca7d70fbf3d249aba4de377ec2a2fa1b98/plugins/com.redhat.ceylon.eclipse.ui/src/com/redhat/ceylon/eclipse/code/navigator/CeylonNavigatorContentProvider.java#L139),

The reason is roughly always the same : we search for a package that according to the name of the module. However if a Ceylon has been overridden by another Ceylon module (with a different name), this will not work. There we should  use the real module name (`ArtifactResult.getName()`) that should be known for such modules.

**Comment by davidfestal** [comment]
@FroMage : do you want me to add build XSD file for the overrides ?

**Comment by quintesse** [comment]
> do you want me to add build XSD file for the overrides ?

If you know how, just do it, I doubt he will complain :)

**Comment by FroMage** [comment]
@davidfestal: I don't understand how you're having issues here. In theory the CMR handles the overrides and so you never get to see the original import that has been replaced and so there should never be a `Module` loaded for the one we're replacing. In theory also, overrides are not applied to the module descriptors we're compiling, so if the typechecker creates a `Module` from a source `ModuleImport` the override should not be applied.

Do you have a test I can use to understand better?

**Comment by davidfestal** [comment]
Try to import `ceylon.collection` version `1.1.3` (which will not be found), and, in the `overrides.xml` file, override it with the 1.1.0 version (using a `set` tag).

Then debug the build of this project, and, just after the call to `ModuleValidator.verifyModuleDependencyTree()`, take a look at the list of modules of the Ceylon context (`Context.getModules().getListOfModules()`).

 You should find out that the `ceylon.collection` module that is here still has the `1.1.3` version.

It seems quite natural to me, since the `overrides` file only changes the artifact result that is returned by the CMR during for a module import during module resolution. But the name and version of the model Module object itself (created *before* the resolution) is not changed.

**Comment by davidfestal** [comment]
@FroMage : is my scenario clear enough to explain the situation, or do you want me to create a compiler test for this ?

**Comment by FroMage** [comment]
OK, I get it, that's precisely what I do not support: overriding of imports where we're actuallying compiling the module descriptors. I thought that'd be counter-intuitive to override something you're actually compiling. Overrides are more meant to fix things you have no control over. So I guess I should make it so that the CMR returns the `1.1.3` version you asked for in this case, right?

**Comment by davidfestal** [comment]
@gavinking : No more error popups if the xml is invalid. But an error marker, that can even report you to the invalid line as reported by @FroMage Overrides parsing.
More if no `overrides` file exist, you can create it from the properties page, and it is created with the right XML schema namespace.
I also added a XML catalog contribution and a documented XSD resource.
So when opened in the standard EclipseXML Editor (of WTP), the created XML file can be built with schema-aware completion, and documentation in hovers and in the standard XMLdocumentation view.

**Comment by gavinking** [comment]
Excellent! Sounds great!

**Comment by davidfestal** [comment]
@FroMage : you are certainly right, but I wonder if we couldn't still have some problems with transitively imported Ceylon modules that would be overridden, when trying to navigate inside their source.
I have to take some more time to submit a more detailed use case to you.

**Comment by davidfestal** [comment]
@FroMage  : so do you mean that *only* imports that are *not visible at all* from source modules can be overriden ? In other words, that modules that are *transitively imported* (thanks the the export mechanism) by some source module cannot be overridden ?
In this case, yes, we won't have any problem.

**Comment by FroMage** [comment]
Every module loaded from binary has overrides applied. No module loaded from source has overrides applied. Now, this is a decision based on 1/ laziness because it would involve adapting the `ModuleVisitor` and 2/ not applying overrides to things the user has control over (module descriptor sources) and 3/ so that the module descriptor which is compiled is exactly the one from the source (and not a mix of source+overrides).

**Comment by davidfestal** [comment]
Yes, but the problem is the, according to how the module validation and resolving is implemented currently, loading different archive for modules without updating the module descriptor leads to an inconsistent situation and bugs, even for binary-loaded modules, as shown by the following test : https://github.com/ceylon/ceylon-compiler/blob/92831469fef1dad2e9a3f7ff8bb6df992846ed31/test/src/com/redhat/ceylon/compiler/java/test/cmr/CMRTests.java#L1509

**Comment by davidfestal** [comment]
I'm working on it.
